### DISCUSSION TITLE
Adding types to a recursive generator (yield, Unkown)

### DISCUSSION THREAD

**Discussion by FrancescElies** [microsoft/pyright#4817] [Q&A] [answered]
Given this example, I have two questions
```python
NumbersDict = dict[str, int | float | "NumbersDict"]
# NOTE: the following line fails at runtime, but this has nothing to do with pyright
# NumbersDict = dict[str, int | float | "NumbersDict"]


def give_ints(container: NumbersDict | int | float):
    match container:
        case dict():
            yield from (give_ints(x) for x in container.values())
        case int():
            yield container
        case _:
            yield -1


print([x for x in give_ints({"a": {"aa": {"aaa": 1}}, "b": 2, "c": {"cc": 3}})])
reveal_type(give_ints)  #  Generator[Unknown | int, None, None]"
```

- `NumbersDict = dict[str, int | float | "NumbersDict"]` is not valid python code, do I see that right? (This has nothing to do with pyright but just taking the change to ask)
- Why the type of `(give_ints(x) for x in container.values())` is `(Generator[Unknown, None, None])`?
  What would be the correct way to add types to such a function?
  
  Thanks in advance for your reply
  
  PS: I used python 3.11 and lastest pyright v1.1.300 (22-03-2023)

**Comment by erictraut** [comment, marked as answer]
Yes, PEP 604 (which introduced the `|` form for unions) has some runtime limitations. These are being fixed over time, but Python 3.11 still has some of these limitations in place. You can avoid this by using the older `typing.Union` form in cases where the new form causes a runtime error.

The problem in your code is that sometimes you're yielding an `int` object, but other times you're yielding a `Generator` object. To have consistent typing, you would need to always yield an `int` object, like this:

```python
NumbersDict = dict[str, Union[int, float, "NumbersDict"]]

def give_ints(container: Union[NumbersDict, int, float]):
    match container:
        case dict():
            for y in (give_ints(x) for x in container.values()):
                yield from y
        case int():
            yield container
        case _:
            yield -1
```

**Reply by FrancescElies** [reply]
Thanks for the insight
